src/libraries/System.Linq/src/System/Linq/Count.cs

```diff
@@ -72,6 +72,59 @@ public static int Count<TSource>(this IEnumerable<TSource> source, Func<TSource,
             return count;
         }
 
+        /// <summary>
+        ///   Attempts to determine the number of elements in a sequence without forcing an enumeration.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source" />.</typeparam>
+        /// <param name="source">A sequence that contains elements to be counted.</param>
+        /// <param name="count">
+        ///     When this method returns, contains the count of <paramref name="source" /> if successful,
+        ///     or zero if the method failed to determine the count.</param>
+        /// <returns>
+        ///   <see langword="true" /> if the count of <paramref name="source"/> can be determined without enumeration;
+        ///   otherwise, <see langword="false" />.
+        /// </returns>
+        /// <remarks>
+        ///   The method performs a series of type tests, identifying common subtypes whose
+        ///   count can be determined without enumerating; this includes <see cref="ICollection{T}"/>,
+        ///   <see cref="ICollection"/> as well as internal types used in the LINQ implementation.
+        ///
+        ///   The method is typically a constant-time operation, but ultimately this depends on the complexity
+        ///   characteristics of the underlying collection implementation.
+        /// </remarks>
+        public static bool TryGetNonEnumeratedCount<TSource>(this IEnumerable<TSource> source, out int count)
```

Comment:
This doesn't need to be generic?

Reply:
We could do that, however it becomes more difficult to check for generic interfaces, which would require some form of reflection.

Reply:
>which would require some form of reflection. 

That's definitely a no-go because the point of using this is performance. Perhaps those generic interfaces need a non-generic base with `Count` prop. IMO `TryGetNonEnumeratedCount` shouldn't care about `TSource`.